src/shiftedNormL0Box.jl

```diff
@@ -0,0 +1,115 @@
+export ShiftedNormL0Box
+
+mutable struct ShiftedNormL0Box{
+  R <: Real,
+  V0 <: AbstractVector{R},
+  V1 <: AbstractVector{R},
+  V2 <: AbstractVector{R},
+  V3,
+  V4
+} <: ShiftedProximableFunction
+  h::NormL0{R}
+  xk::V0
+  sj::V1
+  sol::V2
+  l::V3
+  u::V4
+  shifted_twice::Bool
+
+  function ShiftedNormL0Box(
+    h::NormL0{R},
+    xk::AbstractVector{R},
+    sj::AbstractVector{R},
+    l,
+    u,
+    shifted_twice::Bool
+  ) where {R <: Real}
+    sol = similar(xk)
+    if sum(l .> u) > 0
+      error("Error: at least one lower bound is greater than the upper bound.")
+    end
+    new{R, typeof(xk), typeof(sj), typeof(sol), typeof(l), typeof(u)}(h, xk, sj, sol, l, u, shifted_twice)
+  end
+
+end
+
+shifted(h::NormL0{R}, xk::AbstractVector{R}, l, u) where {R <: Real} =
+  ShiftedNormL0Box(h, xk, zero(xk), l, u, false)
+
+shifted(
+  ψ::ShiftedNormL0Box{R, V0, V1, V2, V3, V4},
+  sj::AbstractVector{R},
+) where {R <: Real, V0 <: AbstractVector{R}, V1 <: AbstractVector{R}, V2 <: AbstractVector{R}, V3, V4} =
+  ShiftedNormL0Box(ψ.h, ψ.xk, sj, ψ.l, ψ.u, true)
+
+fun_name(ψ::ShiftedNormL0Box) = "shifted L0 pseudo-norm with box indicator"
+fun_expr(ψ::ShiftedNormL0Box) = "t ↦ λ ‖xk + sj + t‖₀ + χ({sj + t .∈ [l,u]})"
+fun_params(ψ::ShiftedNormL0Box) =
+  "xk = $(ψ.xk)\n" * " "^14 * "sj = $(ψ.sj)\n" * " "^14 * "lb = $(ψ.l)\n" * " "^14 * "ub = $(ψ.u)"
+
+
+function prox!(
+  y::AbstractVector{R},
+  ψ::ShiftedNormL0Box{R, V0, V1, V2, V3, V4},
+  q::AbstractVector{R},
+  σ::R,
+) where {R <: Real, V0 <: AbstractVector{R}, V1 <: AbstractVector{R}, V2 <: AbstractVector{R}, V3, V4}
+  
+  c2 = 2 * ψ.λ * σ
+
+  for i ∈ eachindex(q)
+
+    if isa(ψ.l, Real) && isa(ψ.u, Real)
+      li = ψ.l
+      ui = ψ.u
+    elseif isa(ψ.l, Real)
+      li = ψ.l
+      ui = ψ.u[i]
+    elseif isa(ψ.u, Real)
+      li = ψ.l[i]
+      ui = ψ.u
+    else
+      li = ψ.l[i]
+      ui = ψ.u[i]
+    end 
```

Comment:
Maybe that could all be
```julia
li = isa(ψ.l, Real) ? ψ.l : ψ.l[i]
ui = isa(ψ.u, Real) ? ψ.u : ψ.u[i]
```

Reply:
Very good idea ! That is way more concise, I correct this.